pkg/controllers/customcluster_upgrade.go

```diff
@@ -0,0 +1,198 @@
+			log.Error(err, "failed to update the kubeVersion of configmap cluster-config after upgrading")
+			return ctrl.Result{}, err
+		}
+		// Restore the workerPod's status to "provisioned" after upgrading.
+		log.Info("phase changes", "prevPhase", customCluster.Status.Phase, "currentPhase", v1alpha1.ProvisionedPhase)
+		customCluster.Status.Phase = v1alpha1.ProvisionedPhase
+
+		// Delete the upgrading worker.
+		if err := r.ensureWorkerPodDeleted(ctx, generateWorkerKey(customCluster, CustomClusterUpgradeAction)); err != nil {
+			log.Error(err, "failed to delete upgrade worker pod", "worker", generateWorkerKey(customCluster, CustomClusterUpgradeAction))
+			return ctrl.Result{}, err
+		}
+		conditions.MarkTrue(customCluster, v1alpha1.UpgradeCondition)
+
+		return ctrl.Result{}, nil
+	}
+
+	// When upgrade worker pod runs failed, the status of customCluster will change into "provisioned". Deleting this error one will trigger the creation of a new upgrade worker pod.
+	if workerPod.Status.Phase == corev1.PodFailed {
+		log.Info("upgrade worker runs failed, customCluster phase changes", "prevPhase", customCluster.Status.Phase, "currentPhase", v1alpha1.ProvisionedPhase)
+		customCluster.Status.Phase = v1alpha1.UnknownPhase
+		conditions.MarkFalse(customCluster, v1alpha1.UpgradeCondition, v1alpha1.UpgradeWorkerRunFailedReason,
+			clusterv1.ConditionSeverityWarning, "upgrade worker run failed %s/%s", customCluster.Namespace, customCluster.Name)
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// generateScaleDownManageCMD generate a kubespray cmd to upgrade cluster to desired kubeVersion.
+func generateUpgradeManageCMD(kubeVersion string) customClusterManageCMD {
+	if len(kubeVersion) == 0 {
+		return ""
+	}
+
+	cmd := string(KubesprayUpgradeCMDPrefix) + " -e kube_version=v" + strings.TrimPrefix(kubeVersion, "v")
+
+	return customClusterManageCMD(cmd)
+}
+
+// isSupportedVersion checks if a desired version is within a specified range of versions.
+func isSupportedVersion(desiredVersion, minVersion, maxVersion string) bool {
+	desiredVersion = strings.TrimPrefix(desiredVersion, "v")
+	minVersion = strings.TrimPrefix(minVersion, "v")
+	maxVersion = strings.TrimPrefix(maxVersion, "v")
+
+	// Parse the version strings using semver package.
+	desired, err := semver.NewVersion(desiredVersion)
+	if err != nil {
+		return false
+	}
+
+	min, err1 := semver.NewVersion(minVersion)
+	if err1 != nil {
+		return false
+	}
+
+	max, err2 := semver.NewVersion(maxVersion)
+	if err2 != nil {
+		return false
+	}
+
+	// Check the desiredVersion is in the correct scope.
+	if desired.Compare(*min) < 0 || desired.Compare(*max) > 0 {
+		return false
+	}
+
+	return true
+}
+
+// isKubeadmUpgradeSupported check if this upgrading is supported to Kubeadm. kubespray using kubeadm to upgrade, but it is not supported to skip MINOR versions during the upgrade process using Kubeadm.
+func isKubeadmUpgradeSupported(originVersion, targetVersion string) bool {
```

Comment:
can you import this function?

Reply:
There is currently no function that can be called directly. 

the check of cluster-api:  https://github.com/kubernetes-sigs/cluster-api/blob/e3ad957784ae4e1bb751dec336d677b619bde49f/internal/webhooks/cluster.go#L326-L340

in the repo of  kubespray and kubeadm, no corresponding function has been found so far.